### ISSUE TITLE
Array method whitelisting is too permissive

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#730]
https://github.com/danielnixon/eslint-plugin-total-functions/issues/100 introduced a small issue.

The array method whitelisting is too permissive. It doesn't consider the type being stored in the array.

Check out this example. The last line should be flagged (like the second last line is) but isn't.

```ts
type MyReadonlyType = { readonly a: string };
type MyMutableType = { a: string };

const readonlyArrayOfReadonly: readonly MyReadonlyType[] = [{ a: "" }] as const;
const readonlyArrayOfMutable: readonly MyMutableType[] = [{ a: "" }];

// This is fine:
const a: readonly MyReadonlyType[] = readonlyArrayOfReadonly;

// This is fine too:
const b: readonly MyMutableType[] = readonlyArrayOfMutable;

// But this is rightly forbidden:
const c: readonly MyReadonlyType[] = readonlyArrayOfMutable;

// However, we can sneakily achieve that unsafe assignment like so:
const d: readonly MyReadonlyType[] = readonlyArrayOfMutable.concat({ a: "" });
```

There are **no** lint errors in the following (but there should be):
```ts
type MyReadonlyType = { readonly a: string };
type MyMutableType = { a: string };

const mutableVal: MyMutableType = { a: "" };
const readonlyArrayOfMutable: readonly MyMutableType[] = [mutableVal] as const;

const shouldBeImmutable: readonly MyReadonlyType[] =
  readonlyArrayOfMutable.concat({
    a: "",
  });

console.log(shouldBeImmutable[0]);

readonlyArrayOfMutable[0]!.a = "whoops";

// Mutation observed in what should be immutable
console.log(shouldBeImmutable[0]);
```
